src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,245 @@
+  }
+
+  /**
+   * Dispose thumbnail loader.
+   * @returns {void}
+   */
+  dispose(): void {
+    disposeSourceBuffer(this._videoElement);
+    this._currentJob?.stop();
+  }
+
+  private _setTime = (time: number,
+                      videoElement: HTMLVideoElement
+  ): Promise<unknown> => {
+
+    for (let i = 0; i < videoElement.buffered.length; i++) {
+      if (videoElement.buffered.start(i) <= time &&
+          videoElement.buffered.end(i) >= time) {
+        this._videoElement.currentTime = time;
+        log.debug("VTL: Thumbnail already loaded.");
+        return PPromise.resolve(time);
+      }
+    }
+
+    const manifest = this._player.getManifest();
+    if (manifest === null) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_MANIFEST",
+                                                           "No manifest available."));
+    }
+    const contentInfos = getContentInfos(time, manifest);
+    if (contentInfos === null) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_TRACK",
+                                                           "Couldn't find track for this time."));
+    }
+    const initURL = contentInfos.representation.index.getInitSegment()?.mediaURL ?? "";
+    if (initURL === "") {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_INIT_DATA", "No init data for track."));
+    }
+    const segment = getNeededSegment(contentInfos,
+                                     time);
+    if (segment === undefined) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_THUMBNAILS", "Couldn't find thumbnail."));
+    }
+
+    log.debug("VTL: Found thumbnail for time", time, segment);
+
+    if (this._currentJob === undefined) {
+      return this.startJob(contentInfos, time, segment);
+    }
+
+    if (this._currentJob.contentInfos.representation.getMimeTypeString() !==
+        contentInfos.representation.getMimeTypeString()) {
+      this._currentJob.stop();
+      return this.startJob(contentInfos, time, segment);
+    }
+    if (this._currentJob.segment.time !== segment.time ||
+        this._currentJob.segment.duration !== segment.duration ||
+        this._currentJob.segment.mediaURL !== segment.mediaURL) {
+      this._currentJob.stop();
+      return this.startJob(contentInfos, time, segment);
+    }
+
+    // If we reach this endpoint, it means the current job is already handling
+    // the loading for the wanted time (same thumbnail).
+    return this._currentJob.jobPromise;
+  }
+
+  private startJob = (contentInfos: IContentInfos,
+                      time: number,
+                      segment: ISegment
+  ): Promise<unknown> => {
+    const killJob$ = new ReplaySubject();
+
+    const abortError$ = killJob$.pipe(
+      map(() => {
+        throw new VideoThumbnailLoaderError("ABORTED",
+                                            "VideoThumbnailLoaderError: Aborted job.");
+      })
+    );
+
+    const jobPromise = observableMerge(
```

Comment:
Isn't what we technically want more of a `race`?

Reply:
I will check that